### DISCUSSION TITLE
Better documentation for Action and Func fields

### DISCUSSION THREAD

**Discussion by truegoodwill** [dotnet/csharplang#4989] [Language Ideas] [answered]
It's becoming very common to see libraries provide apis with `Action` and `Func` fields that have unexplained parameters.

```csharp
// Example of something I might find in a library (or write in my own library)

public Task<ISubscription> StartSubscription(SubscriptionOptions options)
{
  // ...
}

public sealed class SubscriptionOptions
{
  /// <summary>
  /// Called when the subscription has received a new event message.
  /// </summary>
  public Func<ISubscription, ResolvedEvent, bool, CancellationToken, Task> Event { get; init; }
}
```

All that is fine enough at first ... the parameter types are fairly self-explanatory ... and then ... "OH, WHAT THE HECK IS THE `bool` parameter supposed to represent??"

It would be really nice if we could document it further with something like: 

```csharp
  /// <summary>
  /// Called when the subscription has received a new event message.
  /// </summary>
  /// <param isRealtime>False if the event is read from history. True if the event is streaming in realtime.</param>
  public Func<ISubscription subscription, ResolvedEvent @event, bool isRealtime, CancellationToken cancellationToken, Task> Event { get; init; }
```

I have worked around this in the past to make things more readable for library consumers by either of two methods below:

1. Declared the `Func` callback as a method accepting just a single parameter which is a struct with named and commented fields. The struct was created just for the commenting I could attach to its fields.
2. Declared named delegates, which also provide commenting for each field, though the user has to click around a bit to get to it.

**Comment by ufcpp** [comment, marked as answer]
Similar to https://github.com/dotnet/csharplang/discussions/1883 https://github.com/dotnet/csharplang/discussions/269

**Reply by truegoodwill** [reply]
Thank you
